src/rust/lib/graph-editor/src/component/visualization/fullscreen.rs

```diff
@@ -0,0 +1,338 @@
+    /// phase, or already in fullscreen mode.
+    pub fn enable_fullscreen(&self, target:T, scene:Scene) {
+        if !self.is_fullscreen() {
+            self.transition_to_fullscreen(target, scene)
+        }
+    }
+
+    /// Disables fullscreen mode for the given component. Does nothing if we are in an animation
+    /// phase, or not in fullscreen mode.
+    pub fn disable_fullscreen(&self) {
+        let fullscreen_data = {
+            let state = self.state.borrow();
+            let state = state.deref();
+            if let StateModel::Fullscreen {data, ..} = state {
+                Some(data.clone())
+            } else {
+                None
+            }
+        };
+        if let Some(data) = fullscreen_data {
+            self.transition_to_non_fullscreen(&data)
+        }
+    }
+
+    /// Start the transition to non-fullscreen mode. Triggers the UI component state change and
+    /// starts the animation.
+    fn transition_to_non_fullscreen(&self, source_state:&FullscreenStateData<T>) {
+        let animation_data = source_state.prepare_non_fullscreen_animation();
+        let new_state      = StateModel::TransitioningFromFullscreen { animation_data };
+        self.state.replace(new_state);
+        self.animation.set_position(0.0);
+        self.animation.set_target_position(1.0);
+    }
+
+    /// Start the transition to fullscreen mode. Triggers the UI component state change and
+    /// starts the animation.
+    fn transition_to_fullscreen(&self,target:T, scene:Scene) {
+        let target_state   = FullscreenStateData::new(target, scene);
+        let animation_data = target_state.prepare_fullscreen_animation();
+        let new_state      = StateModel::TransitioningToFullscreen { animation_data,target_state };
+        self.state.replace(new_state);
+        self.animation.set_position(0.0);
+        self.animation.set_target_position(1.0);
+    }
+
+    /// Return a ref clone of the fullscreen element.
+    pub fn get_element(&self) -> Option<T> {
+        match self.state.borrow().deref() {
+            StateModel::Fullscreen{ data, .. } => Some(data.target.clone_ref()),
+            _                              => None,
+        }
+    }
+}
+
+
+
+// ============================
+// === FullscreenStateData  ===
+// ============================
+
+/// The `FullscreenStateData` preserves the initial state of a UI element and provides functionality
+/// to transition the UI component from/to fullscreen state. It handles the direct interactions with
+/// the UI component (setting of size/layers) outside of the animation and and provides the target
+/// state for the animation.
+#[derive(Debug,Clone)]
+pub struct FullscreenStateData<T> {
+    target            : T,
+    scene             : Scene,
+    size_original     : Vector3<f32>,
+    position_original : Vector3<f32>,
+    parent_original   : Option<display::object::Instance>,
+}
```

Comment:
I want to review it again after the animation will be coded in frp. I believe it will be simplified drastically.